pkg/azuredisk/controllerserver.go

```diff
@@ -518,51 +518,70 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
 		}
 	}
 
-	disks, derr := d.cloud.DisksClient.ListByResourceGroup(ctx, d.cloud.ResourceGroup)
-	if derr != nil {
-		return nil, status.Error(codes.Internal, fmt.Sprintf("ListVolumes on rg(%s) failed with error: %v", d.cloud.ResourceGroup, derr.Error()))
+	if d.cloud.KubeClient == nil || d.cloud.KubeClient.CoreV1() == nil || d.cloud.KubeClient.CoreV1().PersistentVolumes() == nil {
+		return nil, status.Errorf(codes.Internal, "ListVolumes failed: no kubeClient, corev1, persistentvolume getter found.")
 	}
-
-	if start != 0 && start >= len(disks) {
-		return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
+	pvList, err := d.cloud.KubeClient.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "ListVolumes failed while fetching PersistentVolumes List with error: %v", err.Error())
 	}
 
-	maxEntries := len(disks) - start
-	if req.MaxEntries > 0 && int(req.MaxEntries) < maxEntries {
-		maxEntries = int(req.MaxEntries)
+	if start != 0 && start >= len(pvList.Items) {
+		return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
 	}
-	pageEnd := start + maxEntries
-
+	curEntryIndex := start
 	vmSet := d.cloud.VMSet
 	entries := []*csi.ListVolumesResponse_Entry{}
-	for i := start; i < pageEnd; i++ {
-		d := disks[i]
-		nodeList := []string{}
-
-		// HyperVGeneration property is only setup for os disks. Only the non os disks should be included in the list
-		if d.DiskProperties == nil || d.DiskProperties.HyperVGeneration == "" {
-			if d.ManagedBy != nil {
-				attachedNode, err := vmSet.GetNodeNameByProviderID(*d.ManagedBy)
-				if err != nil {
-					return nil, err
-				}
-				nodeList = append(nodeList, string(attachedNode))
+	for ; curEntryIndex < len(pvList.Items); curEntryIndex++ {
+		if req.MaxEntries > 0 && len(entries) >= int(req.MaxEntries) {
+			break
+		}
+
+		pv := pvList.Items[curEntryIndex]
+		if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == "disk.csi.azure.com" {
+			diskURI := pv.Spec.CSI.VolumeHandle
+			if err := isValidDiskURI(diskURI); err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, "disk URI(%s) is not valid: %v", diskURI, err)
+			}
+			diskName, err := GetDiskName(diskURI)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "could not get disk name from diskURI(%s) with error(%v)", diskURI, err)
+			}
+			resourceGroup, err := GetResourceGroupFromURI(diskURI)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "could not get resource group from diskURI(%s) with error(%v)", diskURI, err)
+			}
+			disk, derr := d.cloud.DisksClient.Get(ctx, resourceGroup, diskName)
```

Comment:
if there are 1K disks, there would be 1K get disk calls, that's too expensive, is it possible to get all resource groups and then `ListByResourceGroup`, that will save get disk calls a lot. Also would you move this code to a standalone function?

Reply:
Thank you for pointing that out.